src/codegen/codegen_compatibility_visitor.cpp

```diff
@@ -0,0 +1,142 @@
+/*************************************************************************
+ * Copyright (C) 2018-2019 Blue Brain Project
+ *
+ * This file is part of NMODL distributed under the terms of the GNU
+ * Lesser General Public License. See top-level LICENSE file for details.
+ *************************************************************************/
+
+#include <algorithm>
+#include <iostream>
+#include <memory>
+#include <string>
+
+#include "codegen/codegen_compatibility_visitor.hpp"
+#include "utils/logger.hpp"
+#include "visitors/lookup_visitor.hpp"
+
+namespace nmodl {
+namespace codegen {
+
+using visitor::AstLookupVisitor;
+
+/**
+ * \details Checks all the ast::AstNodeTypes that are not compatible with NMODL code
+ * generation and prints related messages. If there is some kind of incompatibility
+ * stop NMODL code generation.
+ */
+bool CodegenCompatibilityVisitor::find_incompatible_ast_nodes(Ast* node) {
+    std::vector<AstNodeType> incompatible_ast_types = {AstNodeType::SOLVE_BLOCK,
+                                                       AstNodeType::TERMINAL_BLOCK,
+                                                       AstNodeType::PARTIAL_BLOCK,
+                                                       AstNodeType::MATCH_BLOCK,
+                                                       AstNodeType::BA_BLOCK,
+                                                       AstNodeType::CONSTANT_BLOCK,
+                                                       AstNodeType::CONSTRUCTOR_BLOCK,
+                                                       AstNodeType::DESTRUCTOR_BLOCK,
+                                                       AstNodeType::DISCRETE_BLOCK,
+                                                       AstNodeType::FUNCTION_TABLE_BLOCK,
+                                                       AstNodeType::INDEPENDENT_BLOCK,
+                                                       AstNodeType::GLOBAL_VAR,
+                                                       AstNodeType::POINTER_VAR,
+                                                       AstNodeType::BBCORE_POINTER_VAR};
+    incompatible_ast_nodes = AstLookupVisitor().lookup(node, incompatible_ast_types);
+
```

Comment:
I prefer small functions instead of large logic in single function. For example, see below associated comments :